### ISSUE TITLE
Server error when computing L-function zeros of number fields

### ISSUE THREAD

**Issue by AndrewVSutherland** [LMFDB/lmfdb#1691]
The page www.lmfdb.org/L/Zeros/NumberField/8.0.2178981.1/ causes a server error (as does the page www.lmfdb.org/L/NumberField/8.0.2178981.1/, but in this case the error is not reported to the user, it just fails to compute the Z-plot and logs the exception in flasklog).  Similar errors happen on many other pages, including:

www.lmfdb.org/L/Zeros/NumberField/8.2.32941720000.1/
www.lmfdb.org/L/Zeros/NumberField/10.0.1185762483.1/
www.lmfdb.org/L/Zeros/NumberField/10.0.52089208083.1/
www.lmfdb.org/L/Zeros/NumberField/8.0.3211264.1/

This is possibly related to #1687, in that both involve std::bad_alloc exceptions inside sage.libs.lcalc.lcalc_Lfunction, but in this case the error is handled more gracefully (it logs the error in flasklog and does not crash the LMFDB server).

**Comment by AndrewVSutherland** [comment]
This problem is more pervasive than #1687, affecting about 1/3 of all number fields (37 out of 100 randomly chosen fields), including:

http://www.lmfdb.org/L/Zeros/NumberField/2.2.2105897.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.3.48090396.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.1.132713424.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.1.136749644.1/
http://www.lmfdb.org/L/Zeros/NumberField/5.5.30500193199204.1/
http://www.lmfdb.org/L/Zeros/NumberField/9.1.35044743937.1/
http://www.lmfdb.org/L/Zeros/NumberField/2.2.277288.1/
http://www.lmfdb.org/L/Zeros/NumberField/9.3.821824919051661504.33/
http://www.lmfdb.org/L/Zeros/NumberField/5.1.1420419586896.1/
http://www.lmfdb.org/L/Zeros/NumberField/5.5.3004489415647729.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.3.280169328600000.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.1.145983303.1/
http://www.lmfdb.org/L/Zeros/NumberField/9.1.29054491825.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.1.97523156.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.3.266545038400.2/
http://www.lmfdb.org/L/Zeros/NumberField/7.3.42047721.1/
http://www.lmfdb.org/L/Zeros/NumberField/18.4.593038347905528851673088.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.1.82116376.1/
http://www.lmfdb.org/L/Zeros/NumberField/9.5.121921699136.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.1.70724744.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.1.76401595.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.3.121772457.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.3.93327424.1/
http://www.lmfdb.org/L/Zeros/NumberField/5.5.83150197280929.1/
http://www.lmfdb.org/L/Zeros/NumberField/5.5.876320654400.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.1.58612336.1/
http://www.lmfdb.org/L/Zeros/NumberField/6.2.55968200000.2/
http://www.lmfdb.org/L/Zeros/NumberField/9.3.80128345500864.2/
http://www.lmfdb.org/L/Zeros/NumberField/5.5.164513793609.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.1.60636508.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.1.113385752.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.1.70433152.1/
http://www.lmfdb.org/L/Zeros/NumberField/5.5.6582876541565184.1/
http://www.lmfdb.org/L/Zeros/NumberField/5.3.320206014000.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.3.85465869.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.1.169523527.1/
http://www.lmfdb.org/L/Zeros/NumberField/7.1.89665487.1/

**Comment by edgarcosta** [comment]
The plots and the zeros are computed in parallel, and included with a src="".

We could display error messages with javascript.
For the plot, one can  do something like this:
```
<img src="picture1.gif" onerror="this.onerror=null;this.src='missing.gif';"/>
```
for the zeros, we would need to edit


```
<script type="text/javascript">
  $(function() {
      $("#zeros").load("{{ zeroslink|safe }}");
  });
</script>'
```

**Comment by AndrewVSutherland** [comment]
@edgarcosta With the fix for #1692, we already get this for the zeros, see http://beta.lmfdb.org/L/NumberField/16.2.14383898168677573334952601242898232687.4/, for example (whatever text gets returned shows up, even without any change to the html).  But I like the idea of adding a "not available" picture for the plot when it cannot be computed for some reason.

**Comment by AndrewVSutherland** [comment]
I should not that there are actually two different problems causing the server errors for the pages listed [above](https://github.com/LMFDB/lmfdb/issues/1691#issuecomment-231556070).  For a few of them (e.g. http://www.lmfdb.org/L/Zeros/NumberField/2.2.2105897.1/)  computation is simply taking too long and timing out, but for most (e.g. http://www.lmfdb.org/L/Zeros/NumberField/5.5.164513793609.1/) the error occurs immediately because of a bad_alloc SIGABRT.

@edgarcosta This is an example of where it would be useful to be able to distinguish 500's caused by timeouts, which we used to be able to (sometimes) do because the latter got redirected to a different page by the load balancer -- could this be solved by swapping the gunicorn/load balancer timeouts, i.e. make the load balancer timeout slightly shorter than gunicorn's rather than the other way around?

**Comment by edgarcosta** [comment]
gunicorn timeout is now set at 35s.
and loadbalancer 30s.

**Comment by AndrewVSutherland** [comment]
@edgarcosta Thanks.  BTW, Lfunction.html will display a suitable "not available" message if zeroslink or plotlink are not set, and this is something that we should be able to determine at the time Lfunction.html is rendered.  I'm going to amend the fix for #1692 to do this so that it will say the plotlink is not available rather than displaying an empty stub for an image file that is never going to arrive.

**Comment by AndrewVSutherland** [comment]
I ran some more tests on low degree number fields:

- for 100 randomly chosen quadratic fields there were no 500 errors (corresponding to bad_alloc errors in the sage lcalc library), but there were 14 cases where 502 (timeout) errors occured.  This is surprising because website_zeros is supposed to spend more than 10 seconds.
- for 100 randomly chosen cubic fields there were no errors (all 200s)
- for 100 randomly chosen quartic fields there was only one 500 error; it occured on the page http://www.lmfdb.org/L/Zeros/NumberField/4.2.238186250894592000.1/.
- for 100 randomly chosen quntic fields there were 17 500 errors
- for 100 randomly chosen fields of degree > 5 there were 98 500 errors (only 2 worked).

Based on this data, I think the minimum short term fix (which would also address #1687) is to disable the use of the sage lcalc library for all Dedekind zeta functions of degree > 4 by setting sageLfunction to None in Lfunction.py when the Dedekind zeta L-function is created.  Given the non-trivial number of problems for quadratic fields (about 15%), I could see an argument for always disabling it, but I won't do that right now, since at least these problems don't crash sage.

The longer term fix is to put Dedekind zeta L-functions into the database, as we have for Dirichlet L-functions.  I will raise a separate issue for this.

**Comment by AndrewVSutherland** [comment]
Closed via #1701 (see #1700 for longer term fix).

**Comment by edgarcosta** [comment]
>  This is surprising because website_zeros is supposed to spend more than 10 seconds.

even though there is an argument for time allowed, it isn't used at all
